src/plugins/crypto.c

```diff
@@ -962,6 +962,143 @@ gboolean bd_crypto_luks_resize (const gchar *luks_device, guint64 size, GError *
+    progress_id = bd_utils_report_started (msg);
+    g_free (msg);
+
+    ret = crypt_init_by_name (&cd, luks_device);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to initialize device: %s", strerror_l (-ret, c_locale));
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    ret = crypt_suspend (cd, luks_device);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to suspend device: %s", strerror_l (-ret, c_locale));
+        crypt_free (cd);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    crypt_free (cd);
+    bd_utils_report_finished (progress_id, "Completed");
+    return TRUE;
+}
+
+static gboolean luks_resume (const gchar *luks_device, const guint8 *pass_data, gsize data_len, const gchar *key_file, GError **error) {
+    struct crypt_device *cd = NULL;
+    gboolean success = FALSE;
+    gchar *key_buffer = NULL;
+    gsize buf_len = 0;
+    gint ret = 0;
+    guint64 progress_id = 0;
+    gchar *msg = NULL;
+
+    msg = g_strdup_printf ("Started resuming '%s' LUKS device", luks_device);
+    progress_id = bd_utils_report_started (msg);
+    g_free (msg);
+
+    if ((data_len == 0) && !key_file) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_NO_KEY,
+                     "No passphrase nor key file specified, cannot resume.");
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    ret = crypt_init_by_name (&cd, luks_device);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to initialize device: %s", strerror_l(-ret, c_locale));
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    ret = crypt_load (cd, CRYPT_LUKS, NULL);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to load device's parameters: %s", strerror_l(-ret, c_locale));
+        crypt_free (cd);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    if (key_file) {
+        success = g_file_get_contents (key_file, &key_buffer, &buf_len, error);
+        if (!success) {
+            g_prefix_error (error, "Failed to add key file: %s", strerror_l(-ret, c_locale));
+            crypt_free (cd);
+            bd_utils_report_finished (progress_id, (*error)->message);
+            return FALSE;
+        }
+    } else
+        buf_len = data_len;
+
+    ret = crypt_resume_by_passphrase (cd, luks_device, CRYPT_ANY_SLOT,
+                                      key_buffer ? key_buffer : (char*) pass_data,
+                                      buf_len);
+    g_free (key_buffer);
```

Comment:
This should probably also ``memset()`` the buffer to zeroes so that the key data isn't left in the memory.